vpr/src/timing/PreClusterTimingManager.cpp

```diff
@@ -0,0 +1,278 @@
+
+static float approximate_inter_cluster_delay(const t_arch& arch,
+                                             const t_det_routing_arch& routing_arch,
+                                             const std::string& device_layout);
+
+PreClusterTimingManager::PreClusterTimingManager(bool timing_driven,
+                                                 const AtomNetlist& atom_netlist,
+                                                 const AtomLookup& atom_lookup,
+                                                 const Prepacker& prepacker,
+                                                 e_timing_update_type timing_update_type,
+                                                 const t_arch& arch,
+                                                 const t_det_routing_arch& routing_arch,
+                                                 const std::string& device_layout,
+                                                 const t_analysis_opts& analysis_opts) {
+
+    // If the flow is not timing driven, do not initialize any of the timing
+    // objects and set the valid flag to false. This allows this object to be
+    // passed through the VPR flow when timing is turned off.
+    if (!timing_driven) {
+        is_valid_ = false;
+        return;
+    }
+    is_valid_ = true;
+
+    // Approximate the inter-cluster delay
+    // FIXME: This can probably be simplified. It can also be improved using
+    //        AP information.
+    float inter_cluster_net_delay = approximate_inter_cluster_delay(arch, routing_arch, device_layout);
+    VTR_LOG("Using inter-cluster delay: %g\n", inter_cluster_net_delay);
+
+    // Initialize the timing analyzer
+    clustering_delay_calc_ = std::make_shared<PreClusterDelayCalculator>(atom_netlist,
+                                                                         atom_lookup,
+                                                                         inter_cluster_net_delay,
+                                                                         prepacker);
+    timing_info_ = make_setup_timing_info(clustering_delay_calc_, timing_update_type);
+
+    // Calculate the initial timing
+    timing_info_->update();
+
+    // Create the echo file if requested.
+    if (isEchoFileEnabled(E_ECHO_PRE_PACKING_TIMING_GRAPH)) {
+        auto& timing_ctx = g_vpr_ctx.timing();
+        tatum::write_echo(getEchoFileName(E_ECHO_PRE_PACKING_TIMING_GRAPH),
+                          *timing_ctx.graph, *timing_ctx.constraints, *clustering_delay_calc_, timing_info_->analyzer());
+
+        tatum::NodeId debug_tnode = id_or_pin_name_to_tnode(analysis_opts.echo_dot_timing_graph_node);
+        write_setup_timing_graph_dot(getEchoFileName(E_ECHO_PRE_PACKING_TIMING_GRAPH) + std::string(".dot"),
+                                     *timing_info_, debug_tnode);
+    }
+
+    // Write a timing report.
+    // TODO: Is this a good place for this? Or should it be moved to just after
```

Comment:
I'd leave this ... this gives an early timing report with a very simple 'net-type' delay model.